tests/tests_e3/conftest.py

```diff
@@ -164,3 +176,109 @@ def pytest_sessionfinish(session, exitstatus):
         # pytest docs, which is the closest thing to having an actual Python error in
         # test code.
         session.exitstatus = 3
+
+
+class PypiSimulator:
+    PYPI_URL = "https://pypi.org"
+    MATCHER = regex_compile(f"{PYPI_URL}/pypi/(?P<package>.*)/json")
+    DATA_DIR = path_join(dirname(abspath(__file__)), "pypi_data")
+
+    def __init__(self, requests_mock: Any) -> None:
+        self.mocked_download_urls = set()
+        self.requests_mock = requests_mock
+
+        self.requests_mock.stop()
+
+    def download_file(self, name: str, version: str, request: Any, context: Any) -> str:
+        if not isdir(name):
+            mkdir(name)
+
+        with open(path_join(name, "setup.py"), "w") as fd:
+            fd.write("from setuptools import setup, find_packages\n")
+            fd.write(f"setup(name='{name}',\n")
+            fd.write(f"      version='{version}',\n")
+            fd.write("       packages=find_packages())\n")
+
+        mkdir(path_join(name, name))
+
+        with open(path_join(name, name, "__init__.py"), "w") as fd:
+            fd.write(f"# This is package {name}")
+
+        pkg = Wheel.build(source_dir=name, dest_dir=".")
+        assert isfile(pkg.path)
+
+        with open(pkg.path, "rb") as f:
+            result = f.read()
+
+        context.status_code = 200
+        return result
+
+    def get_metadata(self, request: Any, context: Any) -> str:
+        m = self.MATCHER.match(request.url)
+        if not m:
+            context.status_code = 400
+            return {
+                "message": "Bad Request",
+                "exception": "Mocked pypi received an unexpected request",
+                "url": request.url,
+                "traceback": [tmp.strip() for tmp in traceback_format_stack()],
+            }
+
+        package = m.group("package")
+
+        path = path_join(self.DATA_DIR, "json", f"{package}.json")
+        if not isfile(path):
+            context.status_code = 404
+            return {
+                "message": "Not Found",
+                "exception": f"'{package}.json' file not found",
+                "url": request.url,
+                "package": package,
+                "traceback": [tmp.strip() for tmp in traceback_format_stack()],
+            }
+
+        try:
+            with open(path) as json_file:
+                result = json_loads(json_file.read())
+
+            if "releases" not in result:
+                raise Exception("Bad json metadata: 'releases' key not found")
+        except Exception as e:
+            context.status_code = 500
+            return {
```

Comment:
Returning a dict in a method which should return an `str` :)

Reply:
Common error when adding types to a file ignored by mypy :')